### PR TITLE
Unify fastboot serving with ember-cli

### PR DESCRIPTION
**Pull request by kratiahuja**
**TL;DR: fixes https://github.com/ember-fastboot/ember-cli-fastboot/issues/274 . This will unblock one of the FastBoot 1.0 blockers.** 

## Motivation
Currently FastBoot provides users to serve the base page and the assets via the `ember fastboot` command. It primarily spins up an express server that is responsible for serving the request with a fastboot replaced base page and serves other assets from disk. However, this is similar to what `ember cli` provides via `ember serve`. Since we are not the infrastructure provided by ember-cli, development with `ember fastboot` has become very error prone for various reasons. The PR attempts to resolve those development concerns. 

## How it works
ember-cli@2.12.0-beta.1 and above [introduced a public API](https://github.com/ember-cli/rfcs/pull/80) in ember-cli that allows addons to be responsible for serving the base page and/or assets while still be able to consume the other nice features of `ember serve`. It does so by splitting the task of watching the build and serving assets into different middleware. Infrastructures like FastBoot can then hook a middleware between the watcher and server middleware and be responsible for serving the base page request. 

If apps are using `ember-cli@2.12.0-beta.1` and above, they will be able to serve fastboot rendered base page using `ember serve` and continue using other features of `ember serve`. For apps running ember-cli version below 2.12.0-beta.1, they will not be able to use `ember serve` with FastBoot.

With the new serving logic, you can use `curl` with `ember serve`:
`curl 'http://localhost:4200/' -H 'Accept: text/html'` 

This PR does the following:
- [x] Marks `ember-cli-fastboot` addon to run before `broccoli-serve-files` addon. `broccoli-serve-files` is an in-repo addon in ember-cli
- [x] Uses `serverMiddleware` API of `ember-cli` to provide a middleware to use `fastboot-express-middleware` when there is a base request. 
- [x] Added a message when fastboot is serving the base page
- [x] Added deprecation message in `ember fastboot`
- [x] For any other requests other than base page, it forwards the request to the next middleware in the chain.
- [x] Live reloads the FastBoot sandbox with rebuilt app and vendor files whenever there is a rebuild.
- [x] Fixed existing tests
- [x] Added new tests that test the same functionality as `ember fastboot`

## Advantages
With using `ember serve` to have FastBoot serve the initial base page request we get the following features for free:
- Ability to show build errors when the initial page request is made so the development is easier and better
- Not requiring any infrastructure to explicitly build an express app just for serving.
- Ability to have other assets served by `ember cli` as it does for vanilla app
- Ability to use a proxy server for serving any API requests or any other requests using `ember serve --proxy=http://localhost:3000`
- Ability to have live reload in the browser using `ember-cli` live reload functionality.
- Ability to load the built assets from `tmp` directory instead of dist.

## Deprecation of `ember fastboot`
Since this PR does similar to what `ember fastboot --watch --serve-assets` provides, we will be deprecating `ember fastboot` command. 

## Future TODOs
Following are the minor enhancements that I would like to fix on after this PR lands. They do not block this PR.
- [ ] Make FastBoot live reloading much smarter. We do not need to reload the sandbox if there are only CSS changes. Therefore, we could use `fs-tree-diff` to see if any of the app/vendor files have changed on rebuild and only load it sandbox then.
- [ ] Allow developers to provide `sandboxGlobals` and/or custom sandbox class when serving using FastBoot in development. No one seems to be using this so just added a TODO for now. 
- [ ] Provide a way to turn off fastboot serving at runtime without needing to remove `ember-cli-fastboot`

cc: @rwjblue @tomdale @danmcclain @stefanpenner @josemarluedke @arjansingh

### CODE REVIEW

**Comment by kratiahuja** [comment]
This is now ready for review.

**Comment by kratiahuja** [comment]
Note to self:
- [ ] Test an app with containing broccoli-asset-rev and make sure it still uses fingerprinted assets in manifest.

File: index.js [hunk 1]
```diff
@@ -139,12 +142,65 @@ module.exports = {
...
 
-  postBuild: function() {
+  postBuild: function(result) {
     this.emit('postBuild');
+    if (this.fastboot) {
+      // should we reload fastboot if there are only css changes? Seems it maynot be needed.
+      // TODO: we can do a smarter reload here by running fs-tree-diff on files loaded
+      // in sandbox (explore later).
+      this.ui.writeLine('Reloading FastBoot...');
+      this.fastboot.reload({
```
Comment (stefanpenner) [inline index.js]: Is fastboot `reload` synchronous? If it is asynchronous, we may need to do more work to ensure no requests during this period are served.

File: index.js [hunk 2]
```diff
@@ -139,12 +142,65 @@ module.exports = {
...
+            this.fastboot = new FastBoot({
+              distPath: outputPath
+            });
+          }
+
+          var fastbootMiddleware = FastBootExpressMiddleware({
+            fastboot: this.fastboot
+          });
+
+          fastbootMiddleware(req, resp, next);
```
Comment (stefanpenner) [inline index.js]: the conditional on line 160 suggests we only serve `fastbootMiddelware` for the first request, subsequent requests will fallback to the `else` case.

Comment (kratiahuja) [inline index.js]: Yes that is correct. If I understand correctly, FastBoot is only meant for initial requests and that too for the base page request (basically have your app be server side rendered for the initial request). For all other requests (other assets, subsequent transitions) should be defaulted to `ember-cli`'s asset serving. 

Do you think that is a wrong assumption? Is there a usecase that I might have missed?

File: index.js (hunk 1 above)
Comment (rwjblue) [inline index.js]: We can implement the `instrumentation` API (ember-cli@2.13+) and detect the output files that have changed, to see if we need to reload `fastboot` here.

Comment (kratiahuja) [inline index.js]: Agreed. That will eliminate a lot of boiler plate code that I planned to write :)

**Comment by josemarluedke** [comment]
Awesome work here @kratiahuja! I just tried this branch in our application and I got the following error:

```
$ ember s
Livereload server on http://localhost:49153
Serving on http://localhost:4200/

Build successful - 18687ms.



Slowest Nodes (totalTime => 5% )              | Total (avg)
----------------------------------------------+---------------------
EslintValidationFilter (5)                    | 4874ms (974 ms)
TestTransformFilter (2)                       | 3551ms (1775 ms)
Babel (88)                                    | 3349ms (38 ms)
SassCompiler (3)                              | 1745ms (581 ms)
Concat (16)                                   | 1292ms (80 ms)
Funnel (248)                                  | 1244ms (5 ms)

App is being served by FastBoot
TypeError: Path must be a string. Received undefined
    at assertPath (path.js:7:11)
    at Object.join (path.js:1211:7)
    at EmberApp.readPackageJSON (/Users/josemarluedke/Projects/main/node_modules/fastboot/dist/cjs/ember-app.js:309:27)
    at new EmberApp (/Users/josemarluedke/Projects/main/node_modules/fastboot/dist/cjs/ember-app.js:29:27)
    at FastBoot._buildEmberApp (/Users/josemarluedke/Projects/main/node_modules/fastboot/dist/cjs/index.js:98:21)
    at new FastBoot (/Users/josemarluedke/Projects/main/node_modules/fastboot/dist/cjs/index.js:47:14)
    at Class.<anonymous> (/Users/josemarluedke/Projects/main/node_modules/ember-cli-fastboot/index.js:164:29)
    at Layer.handle [as handle_request] (/Users/josemarluedke/Projects/main/node_modules/express/lib/router/layer.js:95:5)
    at trim_prefix (/Users/josemarluedke/Projects/main/node_modules/express/lib/router/index.js:312:13)
    at /Users/josemarluedke/Projects/main/node_modules/express/lib/router/index.js:280:7
GET / 500 5.984 ms - -
GET /favicon.ico 200 1.707 ms - -
```

```
$ ember -v
ember-cli: 2.12.0-beta.1
node: 7.5.0
os: darwin x64
```

**Comment by mfazekas** [comment]
@kratiahuja with this change `ember serve` will be serving fastboot version if i point my browser to http://localhost:4200/. Is there a way to ask for regular non fastboot other than remove `ember-cli-fastboot` from package.json? Usecase is that you see some error with fastboot and want to check if it happens in non fastboot version too.

**Comment by danmcclain** [comment]
Just spitballing: what if we could skip injected middleware via query params

**Comment by kellyselden** [comment]
@mfazekas @danmcclain https://github.com/ember-fastboot/ember-cli-fastboot#double-build-times-and-no-incremental-builds this is how you can run a normal build without removing the ember-cli-fastboot package.

**Comment by danmcclain** [comment]
@kellyselden I was thinking of a way that wouldn't require you to restart the server (and also double build times are going away 🙌 )

File: index.js [hunk 3]
```diff
@@ -139,12 +142,65 @@ module.exports = {
...
+        } else {
+          // forward the request to the next middleware (example other assets, proxy etc)
+          next();
+        }
+      }.bind(this))
```
Comment (kratiahuja) [inline index.js]: Yeah I totally forgot ember-cli allows arrow functions now. Will update this.

**Comment by kratiahuja** [comment]
@mfazekas @danmcclain I agree this is a valid usecase but not scoped to this PR. I want to first get base functionality in and then work on this enhancement. I also like the idea of the query param to switch fastboot serving on/off at run time. I was also thinking of providing it via `.embercli` config file if possible. Mind opening a separate issue for this and I will be happy to address it once this is merged? I would like to evaluate all options on the table and implement the best feasible one.

File: index.js (hunk 3 above)
Comment (kratiahuja) [inline index.js]: Updated to use arrow functions.

File: index.js (hunk 2 above)
Comment (stefanpenner) [inline index.js]: > Do you think that is a wrong assumption? Is there a usecase that I might have missed?

Ya, that isn't quite correct. If a user hard reloads, they should once again get a fastboot rendered page.

Comment (kratiahuja) [inline index.js]: @stefanpenner They will get a fastboot rendered page. FastBoot is just responsible for serving the index.html with the rendered content. I did test both use cases on a test app.  I am not sure I am following what is not correct.

Comment (stefanpenner) [inline index.js]: Your right. for some reason i read line 174 being within the if statement of 161. Not sure how that happened, my bad.

**Comment by kratiahuja** [comment]
Tested this PR with `fastboot-website` app on ember-cli version `2.12.0-beta.1` and `2.11.0` (this is the current one) and both worked as expected with `ember fastboot --serve-assets` and `ember serve` commands.

**Comment by mfazekas** [comment]
We also tested with our app [percy-web](https://github.com/percy/percy-web/pull/69) and it was working great, we needed it for proxy support. (Only issue is `FASTBOOT_DISABLED` is broken with this PR as it disables fb build but this PR tries to serve it, but that should be addressed in #358).

**Comment by kratiahuja** [comment]
@rwjblue could you please review this? 😄

**Comment by josemarluedke** [comment]
Awesome work @kratiahuja! 

I tested this branch in our application as well a brand new ember app and it works as a charm. 🎉 

Looking forward to have this merged and available to users.

**Comment by mirague** [comment]
This looks good!

**Comment by bcardarella** [comment]
Is there anything else blocking this PR?

**Comment by kratiahuja** [comment]
@bcardarella Nope nothing is blocking this PR except someone reviewing it. I'll ping @rwjblue or someone else again after the ember conference.

**Comment by bcardarella** [comment]
Didn't @stefanpenner already approve?

**Comment by kratiahuja** [comment]
@bcardarella Yes but I think @rwjblue also wanted to review before merging. Unfortunately I don't have merge powers so we will have to wait for someone with merge powers to review and merge.

**Comment by kratiahuja** [comment]
Bumping this again. @rwjblue could you please review?

**Comment by rwjblue** [comment]
Only two super minor changes around messaging, then this is good to go...

File: lib/commands/fastboot-build.js [hunk 4]
```diff
@@ -20,7 +20,8 @@ module.exports = {
       project: this.project
     });
 
-    deprecate("Use of ember fastboot:build is deprecated. Please use ember build instead.");
+    deprecate('ember fastboot:build will be deprecated and removed after FastBoot 1.0 is released.');
```
Comment (kratiahuja) [inline lib/commands/fastboot-build.js]: Yeah but I just wanted to update the deprecation to say we will be dropping this command too.

**Comment by kratiahuja** [comment]
@rwjblue This is good to merge now. Travis is 💚

**Comment by rwjblue** [comment]
Thank you for all your hard work on this @kratiahuja!

**Comment by simonihmig** [comment]
Awesome, thanks indeed! 🎉
